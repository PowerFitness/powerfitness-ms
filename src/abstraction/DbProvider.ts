import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
