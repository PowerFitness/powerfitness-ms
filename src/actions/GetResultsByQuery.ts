import { Action } from '../abstraction/Action';
import { ParsedQs } from 'qs';
import { buildWhereAndConditions } from '../util/buildConditions';
import DbProvider from '../abstraction/DbProvider';
import { Result } from '../types/Result'

interface ResultsMap {userUniqueId?:string, date?:string}

export class GetResultsByQuery extends Action<Array<Result>> {
	queryParams: ParsedQs;
	constructor(dbProvider: DbProvider, queryParams: ParsedQs) {
		super(dbProvider);
		this.queryParams = queryParams;
	}
	buildQuery(): string {
		const { userUniqueId, date } = this.queryParams;
		if (Object.keys(this.queryParams).length === 0) {
			throw new Error('Missing criteria')
		}
		const whereMap: ResultsMap = {
			userUniqueId,
			date
		} as ResultsMap;

		return `select * from result where ${buildWhereAndConditions(whereMap)}`
	}
}

export default GetResultsByQuery;
