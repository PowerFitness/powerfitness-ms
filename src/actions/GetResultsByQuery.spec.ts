import DbProvider from '../abstraction/DbProvider';
import { mock } from 'jest-mock-extended';
import GetResultsByQuery from './GetResultsByQuery';

describe('GetResultsByQuery', () => {
	test('buildQuery', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetResultsByQuery = new GetResultsByQuery(dbProvider, { userUniqueId: 'uid',
			date: 'date' });
		expect(action.buildQuery()).toMatchSnapshot();
	})
	test('buildQuery throws an error if no criteria', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetResultsByQuery = new GetResultsByQuery(dbProvider, { });
		expect(() => action.buildQuery()).toThrowError('Missing criteria')
	})
})
