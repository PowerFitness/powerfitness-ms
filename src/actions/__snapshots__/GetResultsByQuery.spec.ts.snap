// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetResultsByQuery buildQuery 1`] = `"select * from result where userUniqueId = 'uid' AND date = 'date'"`;
